src/commonMain/kotlin/org/mongodb/kbson/serialization/BsonEncoder.kt

```diff
@@ -0,0 +1,181 @@
+package org.mongodb.kbson.serialization
+
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.SerializationStrategy
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.StructureKind
+import kotlinx.serialization.encoding.AbstractEncoder
+import kotlinx.serialization.encoding.CompositeEncoder
+import kotlinx.serialization.modules.SerializersModule
+import org.mongodb.kbson.BsonArray
+import org.mongodb.kbson.BsonBinary
+import org.mongodb.kbson.BsonBoolean
+import org.mongodb.kbson.BsonDocument
+import org.mongodb.kbson.BsonDouble
+import org.mongodb.kbson.BsonInt32
+import org.mongodb.kbson.BsonInt64
+import org.mongodb.kbson.BsonNull
+import org.mongodb.kbson.BsonString
+import org.mongodb.kbson.BsonValue
+
+@OptIn(ExperimentalSerializationApi::class)
+internal sealed class BsonEncoder(
+    override val serializersModule: SerializersModule,
+    val nodeConsumer: (BsonValue) -> Unit
+) : AbstractEncoder() {
+    override fun beginStructure(descriptor: SerialDescriptor): CompositeEncoder =
+        when (descriptor.kind) {
+            StructureKind.LIST ->
+                BsonArrayEncoder(serializersModule) {
+                    pushValue(it)
+                }
+            StructureKind.CLASS ->
+                BsonClassEncoder(serializersModule) {
+                    pushValue(it)
+                }
+            StructureKind.MAP ->
+                BsonDocumentEncoder(serializersModule) {
+                    pushValue(it)
+                }
+            StructureKind.OBJECT -> BsonDocumentEncoder(serializersModule) {
+                // Mimics the Json encode behavior of returning an empty map on Kotlin Objects.
+                pushValue(BsonDocument())
+            }
+
+            else -> error("Unsupported descriptor Kind ${descriptor.kind}")
+        }
+
+    override fun endStructure(descriptor: SerialDescriptor) {
+        nodeConsumer(getCurrent())
+    }
+
+    override fun encodeInt(value: Int) {
+        pushValue(BsonInt32(value))
+    }
+
+    override fun encodeString(value: String) {
+        pushValue(BsonString(value))
+    }
+
+    override fun encodeBoolean(value: Boolean) {
+        pushValue(BsonBoolean(value))
+    }
+
+    override fun encodeByte(value: Byte) {
+        pushValue(BsonInt32(value.toInt()))
+    }
+
+    override fun encodeChar(value: Char) {
+        pushValue(BsonString(value.toString()))
+    }
+
+    override fun encodeDouble(value: Double) {
+        pushValue(BsonDouble(value))
+    }
+
+    override fun encodeFloat(value: Float) {
+        pushValue(BsonDouble(value.toDouble()))
+    }
+
+    override fun encodeLong(value: Long) {
+        pushValue(BsonInt64(value))
+    }
+
+    override fun encodeShort(value: Short) {
+        pushValue(BsonInt32(value.toInt()))
```

Comment:
Are we sure these conversions are correct? From a pure type perspective, it makes sense, but I'm not sure if that maps to how users would use the API? I.e. could we risk that people would want to map shorts to int64 because their data happens to have that format?

I know that MongoDB wouldn't care, it could insert whatever we send, but is there a risk we are breaking schema variants? Maybe @rozza have some input.

I guess the question is: If there is a real concern here, do we need to expose some way to configure how Long/Int/Short/Byte is encoded?

Reply:
I copied the [behavior from the `mongo-java-driver`](https://github.com/mongodb/mongo-java-driver/blob/a9ac6c05a9a12bce110d5a9a89cceedd07843349/bson/src/main/org/bson/codecs/ShortCodec.java#L35):

We could see this as a valid base behavior, if required we could configure the encoder to encode to use `BsonInt32` or `BsonInt64`. But I would leave this as a future feature.